optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,571 @@
+            name (`str`, *optional*, defaults to `""`):
+                To keep track of the name of the current module
+        """
+        QuantLinear = dynamically_import_QuantLinear(
+            use_triton=False, desc_act=self.desc_act, group_size=self.group_size
+        )
+        if isinstance(module, QuantLinear):
+            return
+        for attr in dir(module):
+            layer = getattr(module, attr)
+            name1 = name + "." + attr if name != "" else attr
+            if name1 in names:
+                device = get_device(layer)
+                delattr(module, attr)
+                if isinstance(layer, nn.Linear):
+                    in_features = layer.in_features
+                    out_features = layer.out_features
+                elif isinstance(layer, nn.Conv2d):
+                    in_features = layer.in_channels
+                    out_features = layer.out_channels
+                elif isinstance(layer, Conv1D):
+                    in_features = layer.weight.shape[0]
+                    out_features = layer.weight.shape[1]
+                if not (self.desc_act) or self.group_size == -1:
+                    new_layer = QuantLinear(
+                        self.bits, self.group_size, in_features, out_features, True, use_cuda_fp16=self.use_cuda_fp16
+                    )
+                else:
+                    new_layer = QuantLinear(self.bits, self.group_size, in_features, out_features, True)
+                new_layer.device = device
+                setattr(module, attr, new_layer.to(device))
+        for name1, child in module.named_children():
+            self._replace_by_quant_layers(child, names, name + "." + name1 if name != "" else name1)
+
+    @torch.no_grad()
+    def quantize_model(
+        self,
+        model: nn.Module,
+        tokenizer: Any,
+        dataset: Union[List[str], str],
+        batch_size: int = 1,
+        pad_token_id: Optional[int] = None,
+    ):
+        """
+        Quantize the model using the dataset
+
+        Args:
+            model (`nn.Module`):
+                The model to quantize
+            dataset (`Union[List[str],str]`):
+                The dataset used for quantization. You can provide your own dataset in a list of string or just use the original datasets used
+                in the paper ['wikitext2','c4'].
+            tokenizer (`Any`, defaults to `None`):
+                The tokenizer to use in order to prepare the dataset
+            batch_size (`Optional[int]`, *optional*, defaults to `1`):
+                The batch size of the dataset
+            pad_token_id (`Optional[int]`, *optional*, defaults to `None`):
+                The pad token id. Needed to prepare the dataset when `batch_size` > 1.
+
+        Returns:
+            `nn.Module`: The quantized model
+        """
+
+        if not torch.cuda.is_available():
+            raise RuntimeError("No GPU found. A GPU is needed to quantize model.")
```

Comment:
Do we actually need a GPU to quantize the model? I don't have the details right out of my head now, but is it a strong requirement?

Reply:
(Not considering the actual inference process which I understand requires a GPU, maybe quantization/calibration can be done on CPU too?)

Reply:
I don't think it is a strong requirement. I need to test but you can have a look at `fasterquant` method from `GPTQ` class. Looks like it uses `cuda` to speed up some operations (`cholesky`, `cholesky_inverse`, `matmul`) and I can see that he calls `torch.cuda.synchronize()`.